### PR TITLE
store_forward argument for softmax cross entropy loss

### LINKED ISSUE #1067: Add an option to delete forward result in softmax cross entropy
When the number of output labels is large especially in NLP, forward result of softmax cross entropy `y` consumes a lot of memory. I want an option to delete this value after forward computation.

### PR DESCRIPTION
**Pull request by unnonouno**
I added an argument to change behavior of softmax cross entropy. Only when `store_forward` is True, the loss function stores forward computation result.
fix #1067

### CODE REVIEW

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py [hunk 1]
```diff
@@ -16,28 +15,36 @@
...
 
         # Compute expected value
         loss_expect = 0.0
         count = 0
-        for i in six.moves.range(self.x.shape[0]):
-            if self.t[i] == -1:
+        x = numpy.rollaxis(self.x, 1, self.x.ndim).reshape(
+            (self.t.size, self.x.shape[1]))
+        t = self.t.reshape(self.t.size)
+        for i in six.moves.range(x.shape[0]):
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: How about `for ti, xi in six.moves.zip(t, x):`?

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py [hunk 2]
```diff
@@ -16,28 +15,36 @@
...
         loss_value = float(cuda.to_cpu(loss.data))
 
         # Compute expected value
         loss_expect = 0.0
         count = 0
-        for i in six.moves.range(self.x.shape[0]):
-            if self.t[i] == -1:
+        x = numpy.rollaxis(self.x, 1, self.x.ndim).reshape(
+            (self.t.size, self.x.shape[1]))
+        t = self.t.reshape(self.t.size)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: How about using `ravel`?

**Comment by okuta** [comment]
LGTM except two comments.

**Comment by unnonouno** [comment]
I fixed it.

File: chainer/functions/loss/softmax_cross_entropy.py [hunk 3]
```diff
@@ -178,6 +192,9 @@ def softmax_cross_entropy(x, t, use_cudnn=True, normalize=True):
             determines the normalization constant. If true, this function
             normalizes the cross entropy loss across all instances. If else,
             it only normalizes along a batch size.
+        store_forward (bool): When it is ``True``, the function stores result
+            of forward computation to use it on backward computation. It reduces
```
Comment (okuta) [inline chainer/functions/loss/softmax_cross_entropy.py]: Please fix long line.

**Comment by beam2d** [comment]
I think `store_forward` is confusing a little, since it does not cache the final result of forward (i.e., loss). I propose to use `{cache,keep,store}_{prob,score}` as alternatives.

**Comment by unnonouno** [comment]
Humm, I agree with you. How about `cache_score` ?

**Comment by okuta** [comment]
`cache_score` is good.

**Comment by unnonouno** [comment]
I fixed it.

**Comment by okuta** [comment]
LGTM!
